src/Features/CSharp/Portable/GenerateMember/GenerateParameterizedMember/CSharpGenerateDeconstructMethodService.cs

```diff
@@ -35,5 +43,32 @@ protected override bool AreSpecialOptionsActive(SemanticModel semanticModel)
 
         protected override bool IsValidSymbol(ISymbol symbol, SemanticModel semanticModel)
             => CSharpCommonGenerationServiceMethods.IsValidSymbol();
+
+        public override ImmutableArray<IParameterSymbol> TryMakeParameters(SemanticModel semanticModel, SyntaxNode target, CancellationToken cancellationToken)
+        {
+            // For `if (this is C(0, 0))`, we 'll generate `Deconstruct(out int v1, out int v2)`
+            if (target is PositionalPatternClauseSyntax positionalPattern)
+            {
+                var namesBuilder = positionalPattern.Subpatterns.SelectAsArray(sub =>
+                    semanticModel.GenerateNameForExpression(((ConstantPatternSyntax)sub.Pattern).Expression, false, cancellationToken));
+
+                var names = NameGenerator.EnsureUniqueness(namesBuilder);
+
+                var targetTypes = positionalPattern.Subpatterns.SelectAsArray(sub =>
+                    semanticModel.GetTypeInfo(((ConstantPatternSyntax)sub.Pattern).Expression, cancellationToken).Type);
+
+                return names.ZipAsArray(targetTypes, (name, targetType) =>
+                    CodeGenerationSymbolFactory.CreateParameterSymbol(attributes: default, RefKind.Out, isParams: false, targetType, name));
```

Comment:
```suggestion
                var targetTypes = positionalPattern.Subpatterns.SelectAsArray(sub =>
                    semanticModel.GetTypeInfo(((ConstantPatternSyntax)sub.Pattern).Expression, cancellationToken).Type);

                return positionalPattern.Subpatterns.SelectAsArray((sub, i) =>
                    CodeGenerationSymbolFactory.CreateParameterSymbol(attributes: default, RefKind.Out, isParams: false, semanticModel.GetTypeInfo(((ConstantPatternSyntax)sub.Pattern).Expression, cancellationToken).Type), names[i]));
```

Reply:
note: i'm wary about the above code because it assumes it can get a type off of all expressions.  it likely needs to fallback to `object` if it can't.

Reply:
Sorry, a bit confused here. so, since you are calling the same thing to get targetTypes in the return statement, should I get rid of the targetTypes immutable array altogether?

Reply:
> SelectAsArray((sub, i)

@CyrusNajmabadi I don't think we have a helper where the `Func` provides the index.

Nearest helper is:

https://github.com/dotnet/roslyn/blob/19704e6876c4a5a88267af5cf83a4bfe6287d356/src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs#L169

For this, we might use `CancellationToken` as `TArg`, but we'll have to do `Subpatterns.ToImmutableArray().SelectAsArray(..)`. Is that good enough, or should we introduce a new helper that provides the index for `IEnumerable`s to avoid `ToImmutableArray()` call?

Reply:
@CyrusNajmabadi I have not pushed it yet, but I was able to create a helper function that uses an indexer, like you wrote above, and I was able to pass a test for a null exception with Youssef's suggestion to me of null coalescence on the type parameter. If we decide to go with that option.